zipline/assets/assets.py

```diff
@@ -0,0 +1,535 @@
+
+        # Wipe caches before repopulating
+        self.cache = {}
+        self.sym_cache = {}
+        self.fuzzy_match = {}
+
+        counter = 0
+        for identifier in self.metadata:
+            row = self.metadata.retrieve_metadata(identifier=identifier)
+            self.spawn_asset(identifier=identifier, **row)
+            counter += 1
+
+        self.shared_caches.update(
+            {'by_sid': self.cache,
+             'by_symbol': self.sym_cache,
+             'fuzzy_match': {}}
+        )
+
+    def spawn_asset(self, identifier, **kwargs):
+
+        # Check if the identifier is an int
+        if isinstance(identifier, int) and ('sid' not in kwargs.keys()):
+            kwargs['sid'] = identifier
+
+        # Make sure that the sid exists in the kwargs
+        if 'sid' not in kwargs.keys():
+            kwargs['sid'] = self._assign_sid(identifier)
+
+        # If the identifier coming in was a string and there is no defined
+        # symbol yet, set the symbol to the incoming sid
+        if isinstance(identifier, str) and ('symbol' not in kwargs.keys()):
+            kwargs['symbol'] = identifier
+
+        # If the file_name is in the kwargs, it may be the symbol
+        if 'file_name' in kwargs.keys():
+            file_name = kwargs.pop('file_name')
+            if 'symbol' not in kwargs.keys():
+                kwargs['symbol'] = file_name
+
+        # If the company_name is in the kwargs, it may be the asset_name
+        if 'company_name' in kwargs.keys():
+            company_name = kwargs.pop('company_name')
+            if 'asset_name' not in kwargs.keys():
+                kwargs['asset_name'] = company_name
+
+        # If dates are given as nanos, pop them
+        if 'start_date_nano' in kwargs.keys():
+            kwargs['start_date'] = kwargs.pop('start_date_nano')
+        if 'end_date_nano' in kwargs.keys():
+            kwargs['end_date'] = kwargs.pop('end_date_nano')
+        if 'notice_date_nano' in kwargs.keys():
+            kwargs['notice_date'] = kwargs.pop('notice_date_nano')
+        if 'expiration_date_nano' in kwargs.keys():
+            kwargs['expiration_date'] = kwargs.pop('expiration_date_nano')
+
+        # Process dates
+        if 'start_date' in kwargs.keys():
+            kwargs['start_date'] = self.trading_calendar.\
+                canonicalize_datetime(pd.Timestamp(kwargs['start_date']))
+        if 'end_date' in kwargs.keys():
+            kwargs['end_date'] = self.trading_calendar.\
+                canonicalize_datetime(pd.Timestamp(kwargs['end_date']))
+        if 'notice_date' in kwargs.keys():
+            kwargs['notice_date'] = self.trading_calendar.\
+                canonicalize_datetime(pd.Timestamp(kwargs['notice_date']))
+        if 'expiration_date' in kwargs.keys():
+            kwargs['expiration_date'] = self.trading_calendar.\
+                canonicalize_datetime(pd.Timestamp(kwargs['expiration_date']))
+
+        asset_type = kwargs.pop('asset_type', None)
```

Comment:
I use a .pop to remove asset_type from the kwargs to be passed to Asset(**kwargs)
how does this sound:

        # Build an Asset of the appropriate type
        asset_type = kwargs.pop('asset_type', 'equity')
        if asset_type.lower() == 'equity':
            asset = Equity(**kwargs)
        elif asset_type.lower() == 'future':
            asset = Future(**kwargs)
        else:
            raise InvalidAssetType(asset_type=asset_type)